otlptrace panics when array-type resource attributes are used

### Description

When exporting trace spans associated with a resource that has array-type attributes, otlptrace exporters panic (at least otlptracegrpc, but I think otlptracehttp too).

### Environment

- OS: Linux
- Architecture: x86_64
- Go Version: 1.16.5
- opentelemetry-go version: v1.0.0-RC3

### Steps To Reproduce

1. Run this:

```go
package main

import (
        "context"
        "log"

        "go.opentelemetry.io/otel/attribute"
        "go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracegrpc"
        "go.opentelemetry.io/otel/sdk/resource"
        "go.opentelemetry.io/otel/sdk/trace"
)

func main() {
        exporter, err := otlptracegrpc.New(context.Background())
        if err != nil {
                log.Fatal(err)
        }

        tracerProvider := trace.NewTracerProvider(
                trace.WithSyncer(exporter),
                trace.WithResource(
                        resource.NewSchemaless(
                                attribute.StringSlice("string_array", []string{"a", "b"}),
                        ),
                ),
        )

        tracer := tracerProvider.Tracer("test")
        _, span := tracer.Start(context.Background(), "operation")
        span.End()
```

2. Observe panic:

```
$ go run main.go
panic: runtime error: hash of unhashable type []string

goroutine 1 [running]:
type..hash.go.opentelemetry.io/otel/exporters/otlp/otlptrace/internal/tracetransform.ilsKey·1(0xc0002df400, 0x0, 0x0)
        <autogenerated>:1 +0x35
go.opentelemetry.io/otel/exporters/otlp/otlptrace/internal/tracetransform.Spans(0xc0003702a0, 0x1, 0x1, 0x40fd98, 0x10, 0x907160)
        /home/andrew/go/pkg/mod/go.opentelemetry.io/otel/exporters/otlp/otlptrace@v1.0.0-!r!c3/internal/tracetransform/span.go:57 +0x433
go.opentelemetry.io/otel/exporters/otlp/otlptrace.(*Exporter).ExportSpans(0xc000090960, 0xa72670, 0xc0000240f8, 0xc0003702a0, 0x1, 0x1, 0x0, 0x0)
        /home/andrew/go/pkg/mod/go.opentelemetry.io/otel/exporters/otlp/otlptrace@v1.0.0-!r!c3/exporter.go:44 +0x45
go.opentelemetry.io/otel/sdk/trace.(*simpleSpanProcessor).OnEnd(0xc00007f3c0, 0xa78918, 0xc000386000)
        /home/andrew/go/pkg/mod/go.opentelemetry.io/otel/sdk@v1.0.0-!r!c3/trace/simple_span_processor.go:58 +0x142
go.opentelemetry.io/otel/sdk/trace.(*recordingSpan).End(0xc000001e00, 0x0, 0x0, 0x0)
        /home/andrew/go/pkg/mod/go.opentelemetry.io/otel/sdk@v1.0.0-!r!c3/trace/span.go:280 +0x2d7
main.main()
        /tmp/otelgopanic/main.go:30 +0x473
exit status 2
```

### Expected behavior

Should not panic.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -8,6 +8,10 @@ This project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.htm
 
 ## [Unreleased]
 
+### Fixed
+
+- Slice-valued attributes can correctly be used as map keys. (#2223)
+
 ## [1.0.0-RC3] - 2021-09-02
 
 ### Added

--- a/attribute/value.go
+++ b/attribute/value.go
@@ -73,12 +73,7 @@ func BoolValue(v bool) Value {
 
 // BoolSliceValue creates a BOOLSLICE Value.
 func BoolSliceValue(v []bool) Value {
-	cp := make([]bool, len(v))
-	copy(cp, v)
-	return Value{
-		vtype: BOOLSLICE,
-		slice: cp,
-	}
+	return sliceValue(v, BOOLSLICE)
 }
 
 // IntValue creates an INT64 Value.
@@ -88,14 +83,7 @@ func IntValue(v int) Value {
 
 // IntSliceValue creates an INTSLICE Value.
 func IntSliceValue(v []int) Value {
-	cp := make([]int64, 0, len(v))
-	for _, i := range v {
-		cp = append(cp, int64(i))
-	}
-	return Value{
-		vtype: INT64SLICE,
-		slice: cp,
-	}
+	return sliceValue(v, INT64SLICE)
 }
 
 // Int64Value creates an INT64 Value.
@@ -108,12 +96,7 @@ func Int64Value(v int64) Value {
 
 // Int64SliceValue creates an INT64SLICE Value.
 func Int64SliceValue(v []int64) Value {
-	cp := make([]int64, len(v))
-	copy(cp, v)
-	return Value{
-		vtype: INT64SLICE,
-		slice: cp,
-	}
+	return sliceValue(v, INT64SLICE)
 }
 
 // Float64Value creates a FLOAT64 Value.
@@ -126,12 +109,7 @@ func Float64Value(v float64) Value {
 
 // Float64SliceValue creates a FLOAT64SLICE Value.
 func Float64SliceValue(v []float64) Value {
-	cp := make([]float64, len(v))
-	copy(cp, v)
-	return Value{
-		vtype: FLOAT64SLICE,
-		slice: cp,
-	}
+	return sliceValue(v, FLOAT64SLICE)
 }
 
 // StringValue creates a STRING Value.
@@ -144,11 +122,26 @@ func StringValue(v string) Value {
 
 // StringSliceValue creates a STRINGSLICE Value.
 func StringSliceValue(v []string) Value {
-	cp := make([]string, len(v))
-	copy(cp, v)
-	return Value{
-		vtype: STRINGSLICE,
-		slice: cp,
+	return sliceValue(v, STRINGSLICE)
+}
+
+func sliceValue(v interface{}, vtype Type) Value {
+	// get array type regardless of dimensions
+	typ := reflect.TypeOf(v).Elem()
+	kind := typ.Kind()
+	switch kind {
+	case reflect.Bool, reflect.Int, reflect.Int64,
+		reflect.Float64, reflect.String:
+		val := reflect.ValueOf(v)
+		length := val.Len()
+		frozen := reflect.Indirect(reflect.New(reflect.ArrayOf(length, typ)))
+		reflect.Copy(frozen, val)
+		return Value{
+			vtype: vtype,
+			slice: frozen.Interface(),
+		}
+	default:
+		return Value{vtype: INVALID}
 	}
 }
 
@@ -162,23 +155,7 @@ func StringSliceValue(v []string) Value {
 func ArrayValue(v interface{}) Value {
 	switch reflect.TypeOf(v).Kind() {
 	case reflect.Array, reflect.Slice:
-		// get array type regardless of dimensions
-		typ := reflect.TypeOf(v).Elem()
-		kind := typ.Kind()
-		switch kind {
-		case reflect.Bool, reflect.Int, reflect.Int64,
-			reflect.Float64, reflect.String:
-			val := reflect.ValueOf(v)
-			length := val.Len()
-			frozen := reflect.Indirect(reflect.New(reflect.ArrayOf(length, typ)))
-			reflect.Copy(frozen, val)
-			return Value{
-				vtype: ARRAY,
-				slice: frozen.Interface(),
-			}
-		default:
-			return Value{vtype: INVALID}
-		}
+		return sliceValue(v, ARRAY)
 	}
 	return Value{vtype: INVALID}
 }
@@ -197,10 +174,15 @@ func (v Value) AsBool() bool {
 // AsBoolSlice returns the []bool value. Make sure that the Value's type is
 // BOOLSLICE.
 func (v Value) AsBoolSlice() []bool {
-	if s, ok := v.slice.([]bool); ok {
-		return s
+	if v.vtype != BOOLSLICE {
+		return nil
 	}
-	return nil
+	r := []bool{}
+	s := reflect.ValueOf(v.slice)
+	for i := 0; i < s.Len(); i++ {
+		r = append(r, s.Index(i).Bool())
+	}
+	return r
 }
 
 // AsInt64 returns the int64 value. Make sure that the Value's type is
@@ -212,10 +194,15 @@ func (v Value) AsInt64() int64 {
 // AsInt64Slice returns the []int64 value. Make sure that the Value's type is
 // INT64SLICE.
 func (v Value) AsInt64Slice() []int64 {
-	if s, ok := v.slice.([]int64); ok {
-		return s
+	if v.vtype != INT64SLICE {
+		return nil
 	}
-	return nil
+	r := []int64{}
+	s := reflect.ValueOf(v.slice)
+	for i := 0; i < s.Len(); i++ {
+		r = append(r, s.Index(i).Int())
+	}
+	return r
 }
 
 // AsFloat64 returns the float64 value. Make sure that the Value's
@@ -227,10 +214,15 @@ func (v Value) AsFloat64() float64 {
 // AsFloat64Slice returns the []float64 value. Make sure that the Value's type is
 // INT64SLICE.
 func (v Value) AsFloat64Slice() []float64 {
-	if s, ok := v.slice.([]float64); ok {
-		return s
+	if v.vtype != FLOAT64SLICE {
+		return nil
 	}
-	return nil
+	r := []float64{}
+	s := reflect.ValueOf(v.slice)
+	for i := 0; i < s.Len(); i++ {
+		r = append(r, s.Index(i).Float())
+	}
+	return r
 }
 
 // AsString returns the string value. Make sure that the Value's type
@@ -242,10 +234,15 @@ func (v Value) AsString() string {
 // AsStringSlice returns the []string value. Make sure that the Value's type is
 // INT64SLICE.
 func (v Value) AsStringSlice() []string {
-	if s, ok := v.slice.([]string); ok {
-		return s
+	if v.vtype != STRINGSLICE {
+		return nil
 	}
-	return nil
+	r := []string{}
+	s := reflect.ValueOf(v.slice)
+	for i := 0; i < s.Len(); i++ {
+		r = append(r, s.Index(i).String())
+	}
+	return r
 }
 
 // AsArray returns the array Value as an interface{}.

--- a/attribute/value.go
+++ b/attribute/value.go
@@ -73,7 +73,12 @@ func BoolValue(v bool) Value {
 
 // BoolSliceValue creates a BOOLSLICE Value.
 func BoolSliceValue(v []bool) Value {
-	return sliceValue(v, BOOLSLICE)
+	cp := make([]bool, len(v))
+	copy(cp, v)
+	return Value{
+		vtype: BOOLSLICE,
+		slice: &cp,
+	}
 }
 
 // IntValue creates an INT64 Value.
@@ -83,7 +88,14 @@ func IntValue(v int) Value {
 
 // IntSliceValue creates an INTSLICE Value.
 func IntSliceValue(v []int) Value {
-	return sliceValue(v, INT64SLICE)
+	cp := make([]int64, 0, len(v))
+	for _, i := range v {
+		cp = append(cp, int64(i))
+	}
+	return Value{
+		vtype: INT64SLICE,
+		slice: &cp,
+	}
 }
 
 // Int64Value creates an INT64 Value.
@@ -96,7 +108,12 @@ func Int64Value(v int64) Value {
 
 // Int64SliceValue creates an INT64SLICE Value.
 func Int64SliceValue(v []int64) Value {
-	return sliceValue(v, INT64SLICE)
+	cp := make([]int64, len(v))
+	copy(cp, v)
+	return Value{
+		vtype: INT64SLICE,
+		slice: &cp,
+	}
 }
 
 // Float64Value creates a FLOAT64 Value.
@@ -109,7 +126,12 @@ func Float64Value(v float64) Value {
 
 // Float64SliceValue creates a FLOAT64SLICE Value.
 func Float64SliceValue(v []float64) Value {
-	return sliceValue(v, FLOAT64SLICE)
+	cp := make([]float64, len(v))
+	copy(cp, v)
+	return Value{
+		vtype: FLOAT64SLICE,
+		slice: &cp,
+	}
 }
 
 // StringValue creates a STRING Value.
@@ -122,26 +144,11 @@ func StringValue(v string) Value {
 
 // StringSliceValue creates a STRINGSLICE Value.
 func StringSliceValue(v []string) Value {
-	return sliceValue(v, STRINGSLICE)
-}
-
-func sliceValue(v interface{}, vtype Type) Value {
-	// get array type regardless of dimensions
-	typ := reflect.TypeOf(v).Elem()
-	kind := typ.Kind()
-	switch kind {
-	case reflect.Bool, reflect.Int, reflect.Int64,
-		reflect.Float64, reflect.String:
-		val := reflect.ValueOf(v)
-		length := val.Len()
-		frozen := reflect.Indirect(reflect.New(reflect.ArrayOf(length, typ)))
-		reflect.Copy(frozen, val)
-		return Value{
-			vtype: vtype,
-			slice: frozen.Interface(),
-		}
-	default:
-		return Value{vtype: INVALID}
+	cp := make([]string, len(v))
+	copy(cp, v)
+	return Value{
+		vtype: STRINGSLICE,
+		slice: &cp,
 	}
 }
 
@@ -155,7 +162,23 @@ func sliceValue(v interface{}, vtype Type) Value {
 func ArrayValue(v interface{}) Value {
 	switch reflect.TypeOf(v).Kind() {
 	case reflect.Array, reflect.Slice:
-		return sliceValue(v, ARRAY)
+		// get array type regardless of dimensions
+		typ := reflect.TypeOf(v).Elem()
+		kind := typ.Kind()
+		switch kind {
+		case reflect.Bool, reflect.Int, reflect.Int64,
+			reflect.Float64, reflect.String:
+			val := reflect.ValueOf(v)
+			length := val.Len()
+			frozen := reflect.Indirect(reflect.New(reflect.ArrayOf(length, typ)))
+			reflect.Copy(frozen, val)
+			return Value{
+				vtype: ARRAY,
+				slice: frozen.Interface(),
+			}
+		default:
+			return Value{vtype: INVALID}
+		}
 	}
 	return Value{vtype: INVALID}
 }
@@ -174,15 +197,10 @@ func (v Value) AsBool() bool {
 // AsBoolSlice returns the []bool value. Make sure that the Value's type is
 // BOOLSLICE.
 func (v Value) AsBoolSlice() []bool {
-	if v.vtype != BOOLSLICE {
-		return nil
+	if s, ok := v.slice.(*[]bool); ok {
+		return *s
 	}
-	r := []bool{}
-	s := reflect.ValueOf(v.slice)
-	for i := 0; i < s.Len(); i++ {
-		r = append(r, s.Index(i).Bool())
-	}
-	return r
+	return nil
 }
 
 // AsInt64 returns the int64 value. Make sure that the Value's type is
@@ -194,15 +212,10 @@ func (v Value) AsInt64() int64 {
 // AsInt64Slice returns the []int64 value. Make sure that the Value's type is
 // INT64SLICE.
 func (v Value) AsInt64Slice() []int64 {
-	if v.vtype != INT64SLICE {
-		return nil
+	if s, ok := v.slice.(*[]int64); ok {
+		return *s
 	}
-	r := []int64{}
-	s := reflect.ValueOf(v.slice)
-	for i := 0; i < s.Len(); i++ {
-		r = append(r, s.Index(i).Int())
-	}
-	return r
+	return nil
 }
 
 // AsFloat64 returns the float64 value. Make sure that the Value's
@@ -214,15 +227,10 @@ func (v Value) AsFloat64() float64 {
 // AsFloat64Slice returns the []float64 value. Make sure that the Value's type is
 // INT64SLICE.
 func (v Value) AsFloat64Slice() []float64 {
-	if v.vtype != FLOAT64SLICE {
-		return nil
+	if s, ok := v.slice.(*[]float64); ok {
+		return *s
 	}
-	r := []float64{}
-	s := reflect.ValueOf(v.slice)
-	for i := 0; i < s.Len(); i++ {
-		r = append(r, s.Index(i).Float())
-	}
-	return r
+	return nil
 }
 
 // AsString returns the string value. Make sure that the Value's type
@@ -234,15 +242,10 @@ func (v Value) AsString() string {
 // AsStringSlice returns the []string value. Make sure that the Value's type is
 // INT64SLICE.
 func (v Value) AsStringSlice() []string {
-	if v.vtype != STRINGSLICE {
-		return nil
+	if s, ok := v.slice.(*[]string); ok {
+		return *s
 	}
-	r := []string{}
-	s := reflect.ValueOf(v.slice)
-	for i := 0; i < s.Len(); i++ {
-		r = append(r, s.Index(i).String())
-	}
-	return r
+	return nil
 }
 
 // AsArray returns the array Value as an interface{}.

--- a/attribute/value.go
+++ b/attribute/value.go
@@ -285,14 +285,22 @@ func (v Value) AsInterface() interface{} {
 // Emit returns a string representation of Value's data.
 func (v Value) Emit() string {
 	switch v.Type() {
-	case ARRAY, BOOLSLICE, INT64SLICE, FLOAT64SLICE, STRINGSLICE:
+	case ARRAY:
 		return fmt.Sprint(v.slice)
+	case BOOLSLICE:
+		return fmt.Sprint(*(v.slice.(*[]bool)))
 	case BOOL:
 		return strconv.FormatBool(v.AsBool())
+	case INT64SLICE:
+		return fmt.Sprint(*(v.slice.(*[]int64)))
 	case INT64:
 		return strconv.FormatInt(v.AsInt64(), 10)
+	case FLOAT64SLICE:
+		return fmt.Sprint(*(v.slice.(*[]float64)))
 	case FLOAT64:
 		return fmt.Sprint(v.AsFloat64())
+	case STRINGSLICE:
+		return fmt.Sprint(*(v.slice.(*[]string)))
 	case STRING:
 		return v.stringly
 	default:

--- a/exporters/otlp/otlptrace/internal/tracetransform/span_test.go
+++ b/exporters/otlp/otlptrace/internal/tracetransform/span_test.go
@@ -266,7 +266,7 @@ func TestSpanData(t *testing.T) {
 		DroppedAttributes: 1,
 		DroppedEvents:     2,
 		DroppedLinks:      3,
-		Resource:          resource.NewSchemaless(attribute.String("rk1", "rv1"), attribute.Int64("rk2", 5)),
+		Resource:          resource.NewSchemaless(attribute.String("rk1", "rv1"), attribute.Int64("rk2", 5), attribute.StringSlice("rk3", []string{"sv1", "sv2"})),
 		InstrumentationLibrary: instrumentation.Library{
 			Name:    "go.opentelemetry.io/test/otel",
 			Version: "v0.0.1",

